File: hphp/hack/src/typing/typing_env_types.ml
(*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

(* cf: typing_env_types_sig.mli - These files should be the same *)
open Hh_prelude
open Typing_defs

type locl_ty = Typing_defs.locl_ty

type local_id_set_t = Local_id.Set.t

type local_env = {
  per_cont_env: Typing_per_cont_env.t;
  local_using_vars: local_id_set_t;
}

type expr_tree_env = {
  dsl: Aast.class_name;
  outer_locals: Typing_local_types.t;
}

(** See the .mli file for the documentation of fields. *)
type env = {
  expression_id_provider: Expression_id.provider;
  tvar_id_provider: Tvid.provider;
  fresh_typarams: SSet.t;
  lenv: local_env;
  genv: genv;
  decl_env: Decl_env.env;
  in_loop: bool;
  in_try: bool;
  in_lambda: bool;
  in_expr_tree: expr_tree_env option;
  in_macro_splice: Typing_local_types.t option;
  checked: Tast.check_status;
  tracing_info: Decl_counters.tracing_info option;
  tpenv: Type_parameter_env.t;
  log_levels: int SMap.t;
  inference_env: Typing_inference_env.t;
  rank: int;
  check_rank: bool;
  allow_wildcards: bool;
  big_envs: (Pos.t * env) list ref;
  fun_tast_info: Tast.fun_tast_info option;
  emit_string_coercion_error: bool;
}

(** See the .mli file for the documentation of fields. *)
and genv = {
  tcopt: TypecheckerOptions.t;
  callable_pos: Pos.t;
  function_pos: Pos.t;
  readonly: bool;
  return: Typing_env_return_info.t;
  params: (locl_ty * Pos.t * locl_ty option) Local_id.Map.t;
  parent: (string * decl_ty) option;
  self: (string * locl_ty) option;
  static: bool;
  fun_kind: Ast_defs.fun_kind;
  val_kind: Typing_defs.val_kind;
  fun_is_ctor: bool;
  file: Relative_path.t;
  current_module: Ast_defs.id option;
  current_package: Aast_defs.package_membership option;
  soft_package_requirement: Ast_defs.id option;
  this_internal: bool;
  this_support_dynamic_type: bool;
  no_auto_likes: bool;
  needs_concrete: bool;
}

let initial_local tpenv =
  {
    per_cont_env =
      Typing_per_cont_env.(initial_locals { empty_entry with tpenv });
    local_using_vars = Local_id.Set.empty;
  }

let empty ?origin ?(mode = FileInfo.Mstrict) ctx file ~droot =
  {
    expression_id_provider = Expression_id.make_provider ();
    tvar_id_provider = Tvid.make_provider ();
    fresh_typarams = SSet.empty;
    lenv = initial_local Type_parameter_env.empty;
    in_loop = false;
    in_try = false;
    in_lambda = false;
    in_expr_tree = None;
    in_macro_splice = None;
    checked = Tast.COnce;
    decl_env = { Decl_env.mode; droot; droot_member = None; ctx };
    tracing_info =
      Option.map origin ~f:(fun origin -> { Decl_counters.origin; file });
    genv =
      {
        tcopt = Provider_context.get_tcopt ctx;
        callable_pos = Pos.none;
        function_pos = Pos.none;
        readonly = false;
        return =
          {
            (* Actually should get set straight away anyway *)
            Typing_env_return_info.return_type = mk (Reason.none, Tunion []);
            return_disposable = false;
            return_ignore_readonly = false;
          };
        params = Local_id.Map.empty;
        self = None;
        static = false;
        val_kind = Other;
        parent = None;
        fun_kind = Ast_defs.FSync;
        fun_is_ctor = false;
        file;
        current_module = None;
        current_package = None;
        soft_package_requirement = None;
        this_internal = false;
        this_support_dynamic_type = false;
        no_auto_likes = false;
        needs_concrete = false;
      };
    tpenv = Type_parameter_env.empty;
    log_levels = TypecheckerOptions.log_levels (Provider_context.get_tcopt ctx);
    inference_env = Typing_inference_env.empty_inference_env;
    rank = 0;
    check_rank = false;
    allow_wildcards = false;
    big_envs = ref [];
    fun_tast_info = None;
    emit_string_coercion_error = true;
  }

let get_log_level env key =
  Option.value (SMap.find_opt key env.log_levels) ~default:0

let next_cont_opt env =
  Typing_per_cont_env.get_cont_option
    Typing_continuations.Next
    env.lenv.per_cont_env

let get_tpenv env =
  match next_cont_opt env with
  | None -> Type_parameter_env.empty
  | Some entry -> entry.Typing_per_cont_env.tpenv

let get_pos_and_kind_of_generic env name =
  match Type_parameter_env.get_with_pos name (get_tpenv env) with
  | Some r -> Some r
  | None -> Type_parameter_env.get_with_pos name env.tpenv

let get_lower_bounds env name =
  let tpenv = get_tpenv env in
  let local = Type_parameter_env.get_lower_bounds tpenv name in
  let global = Type_parameter_env.get_lower_bounds env.tpenv name in
  Typing_set.union local global

let get_upper_bounds env name =
  let tpenv = get_tpenv env in
  let local = Type_parameter_env.get_upper_bounds tpenv name in
  let global = Type_parameter_env.get_upper_bounds env.tpenv name in
  Typing_set.union local global

(* Get bounds that are both an upper and lower of a given generic *)
let get_equal_bounds env name =
  let lower = get_lower_bounds env name in
  let upper = get_upper_bounds env name in
  Typing_set.inter lower upper

let get_tparams_in_ty_and_acc env acc ty =
  let tparams_visitor env =
    object (this)
      inherit [SSet.t] Type_visitor.locl_type_visitor

      method! on_tgeneric acc _ s =
        (* Not traversing args, although they may contain Tgenerics (higher kinds only) *)
        SSet.add s acc

      method! on_tvar acc r ix =
        let (_env, ty) =
          Typing_inference_env.expand_var env.inference_env r ix
        in
        match get_node ty with
        | Tvar _ -> acc
        | _ -> this#on_type acc ty
    end
  in
  (tparams_visitor env)#on_type acc ty

let get_rank { rank; _ } = rank

let increment_rank env = { env with rank = env.rank + 1; check_rank = true }

let decrement_rank env = { env with rank = env.rank - 1 }

let should_check_rank { check_rank; _ } = check_rank

File: hphp/hack/src/typing/typing_env_types.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.
use crate::typing_env_return_info;
use crate::typing_make_type::TypeBuilder;
use crate::typing_per_cont_env::{PerContEntry, TypingPerContEnv};
use arena_trait::Arena;
use decl_provider_rust::DeclProvider;
use ocamlrep::{Allocator, ToOcamlRep, Value};
use oxidized::pos::Pos as OwnedPos;
use oxidized::ToOxidized;
use oxidized::{relative_path, typechecker_options};
use oxidized_by_ref::pos::Pos;
use std::cell::RefCell;
use std::collections::{BTreeMap, BTreeSet};
use typing_ast_rust::typing_inference_env::InferenceEnv;

pub use typing_collections_rust::*;
pub use typing_defs_rust::typing_logic::{SubtypeProp, *};
pub use typing_defs_rust::{Ty, *};

pub struct Env<'a> {
    pub ident_counter: isize,
    pub function_pos: &'a Pos<'a>,
    pub fresh_typarams: SSet<'a>,
    pub lenv: LocalEnv<'a>,
    pub genv: Genv<'a>,
    pub log_levels: SMap<'a, isize>,
    pub inference_env: InferenceEnv<'a>,
}

impl<'a> Env<'a> {
    pub fn new<'b>(function_pos: &'b OwnedPos, genv: Genv<'a>) -> Self {
        let function_pos = Pos::from_oxidized_in(function_pos, genv.builder.bumpalo());
        Env {
            ident_counter: 0,
            function_pos,
            fresh_typarams: SSet::empty(),
            inference_env: InferenceEnv::new(genv.builder),
            lenv: LocalEnv::initial_local(genv.builder),
            genv,
            log_levels: SMap::empty(),
        }
    }

    pub fn builder(&self) -> &'a TypeBuilder<'a> {
        self.genv.builder
    }

    pub fn bld(&self) -> &'a TypeBuilder<'a> {
        self.builder()
    }

    pub fn provider(&self) -> &'a dyn DeclProvider {
        self.genv.provider
    }

    pub fn ast_pos(&self, pos: &OwnedPos) -> &'a Pos<'a> {
        Pos::from_oxidized_with_file_in(
            pos,
            self.function_pos.filename(),
            self.genv.builder.bumpalo(),
        )
    }

    pub fn set_function_pos(&mut self, pos: &OwnedPos) {
        self.function_pos = Pos::from_oxidized_in(pos, self.genv.builder.bumpalo());
    }

    pub fn set_return_type(&mut self, ty: Ty<'a>) {
        self.genv.return_info.type_.type_ = ty;
    }
}

impl<'a> ToOxidized for Env<'a> {
    type Target = oxidized::typing_env_types::Env;

    fn to_oxidized(&self) -> oxidized::typing_env_types::Env {
        let Env {
            ident_counter: _,
            function_pos,
            fresh_typarams: _,
            lenv,
            genv,
            log_levels: _,
            inference_env,
        } = self;
        oxidized::typing_env_types::Env {
            function_pos: function_pos.to_oxidized(),
            fresh_typarams: BTreeSet::new(), // TODO(hrust)
            lenv: lenv.to_oxidized(),
            genv: genv.to_oxidized(),
            decl_env: (),
            in_loop: false,
            in_try: false,
            in_case: false,
            inside_constructor: false,
            inside_ppl_class: false,
            global_tpenv: oxidized::type_parameter_env::TypeParameterEnv {
                tparams: BTreeMap::new(),
                consistent: true,
            },
            log_levels: BTreeMap::new(), // TODO(hrust)
            inference_env: inference_env.to_oxidized(),
            allow_wildcards: false,
            big_envs: RefCell::new(vec![]),
            pessimize: false,
        }
    }
}

impl ToOcamlRep for Env<'_> {
    fn to_ocamlrep<'a, A: Allocator>(&self, alloc: &'a A) -> Value<'a> {
        self.to_oxidized().to_ocamlrep(alloc)
    }
}

pub struct Genv<'a> {
    pub tcopt: typechecker_options::TypecheckerOptions,
    pub return_info: typing_env_return_info::TypingEnvReturnInfo<'a>,
    pub params: Map<'a, LocalId<'a>, (Ty<'a>, ParamMode)>,
    pub file: relative_path::RelativePath,
    pub builder: &'a TypeBuilder<'a>,
    pub provider: &'a dyn DeclProvider,
}

impl<'a> ToOxidized for Genv<'a> {
    type Target = oxidized::typing_env_types::Genv;

    fn to_oxidized(&self) -> oxidized::typing_env_types::Genv {
        // TODO(hrust) most fields of Genv
        oxidized::typing_env_types::Genv {
            tcopt: oxidized::global_options::GlobalOptions::default(),
            return_: oxidized::typing_env_return_info::TypingEnvReturnInfo {
                type_: oxidized::typing_defs::PossiblyEnforcedTy {
                    enforced: true,
                    type_: oxidized::typing_defs_core::Ty(
                        oxidized::typing_reason::Reason::Rnone,
                        Box::new(oxidized::typing_defs_core::Ty_::Tmixed),
                    ),
                },
                disposable: false,
                mutable: false,
                explicit: true,
                void_to_rx: false,
            },
            params: BTreeMap::new(),
            condition_types: oxidized::s_map::SMap::new(),
            parent: None,
            self_: None,
            static_: false,
            fun_kind: oxidized::ast_defs::FunKind::FSync,
            val_kind: oxidized::typing_defs::ValKind::Lval,
            fun_mutable: None,
            file: oxidized::relative_path::RelativePath::make(
                oxidized::relative_path::Prefix::Root,
                std::path::PathBuf::new(),
            ),
        }
    }
}

#[derive(Debug)]
pub struct LocalEnv<'a> {
    pub per_cont_env: TypingPerContEnv<'a>,
    // TODO(hrust): mutability
    // TODO(hrust): reactivity
    // TODO(hrust): using vars
}

impl<'a> LocalEnv<'a> {
    pub fn initial_local<A: Arena>(arena: &'a A) -> Self {
        LocalEnv {
            per_cont_env: TypingPerContEnv::initial_locals(arena, PerContEntry::empty()),
        }
    }
}

impl<'a> ToOxidized for LocalEnv<'a> {
    type Target = oxidized::typing_env_types::LocalEnv;

    fn to_oxidized(&self) -> oxidized::typing_env_types::LocalEnv {
        let LocalEnv { per_cont_env } = self;
        oxidized::typing_env_types::LocalEnv {
            per_cont_env: per_cont_env.to_oxidized(),
            local_mutability: BTreeMap::new(),
            local_reactive: oxidized::typing_env_types::Reactivity::Nonreactive,
            local_using_vars: BTreeSet::new(),
        }
    }
}

pub use oxidized_by_ref::local_id::LocalId;
